extensions/socket/internal.m

```diff
@@ -0,0 +1,657 @@
+    LuaSkin *skin = [LuaSkin shared];
+    [skin checkArgs:LS_TUSERDATA, USERDATA_TAG, LS_TBREAK];
+    HSAsyncSocket* asyncSocket = getUserData(L, 1);
+
+    NSDictionary *info = @{
+        @"connectedAddress" : asyncSocket.connectedAddress ?: @"",
+        @"connectedHost" : asyncSocket.connectedHost ?: @"",
+        @"connectedPort" : @(asyncSocket.connectedPort),
+        @"isConnected": @(asyncSocket.isConnected),
+        @"isDisconnected": @(asyncSocket.isDisconnected),
+        @"isIPv4": @(asyncSocket.isIPv4),
+        @"isIPv4Enabled": @(asyncSocket.isIPv4Enabled),
+        @"isIPv4PreferredOverIPv6": @(asyncSocket.isIPv4PreferredOverIPv6),
+        @"isIPv6": @(asyncSocket.isIPv6),
+        @"isIPv6Enabled": @(asyncSocket.isIPv6Enabled),
+        @"isSecure": @(asyncSocket.isSecure),
+        @"localAddress" : asyncSocket.localAddress ?: @"",
+        @"localHost" : asyncSocket.localHost ?: @"",
+        @"localPort" : @(asyncSocket.localPort),
+        @"timeout" : @(asyncSocket.timeout),
+        @"userData" : asyncSocket.userData ?: @"",
+    };
+
+    [skin pushNSObject:info];
+    return 1;
+}
+
+
+static int userdata_gc(lua_State *L) {
+    asyncSocketUserData *userData = lua_touserdata(L, 1);
+    HSAsyncSocket* asyncSocket = (__bridge_transfer HSAsyncSocket *)userData->asyncSocket;
+    userData->asyncSocket = nil;
+
+    [asyncSocket disconnect];
+    [asyncSocket setDelegate:nil delegateQueue:NULL];
+    asyncSocket.readCallback = [[LuaSkin shared] luaUnref:refTable ref:asyncSocket.readCallback];
+    asyncSocket = nil;
+
+    return 0;
+}
+
+static int userdata_tostring(lua_State* L) {
+    HSAsyncSocket* asyncSocket = getUserData(L, 1);
+
+    BOOL isServer = (asyncSocket.userData == SERVER) ? true : false;
+    NSString *theHost = isServer ? asyncSocket.localHost : asyncSocket.connectedHost;
+    UInt16 thePort = isServer ? asyncSocket.localPort : asyncSocket.connectedPort;
+
+    lua_pushstring(L, [[NSString stringWithFormat:@"%s: %@:%hu (%p)", USERDATA_TAG, theHost, thePort, lua_topointer(L, 1)] UTF8String]);
+    return 1;
+}
+
+static const luaL_Reg socketLib[] = {
+    {"new", socket_new},
+    {"parseAddress", socket_parseAddress},
+
+    {NULL, NULL} // This must end with an empty struct
+};
+
+static const luaL_Reg socketObjectLib[] = {
+    {"connect", socket_connect},
+    {"listen", socket_listen},
+    {"disconnect", socket_disconnect},
+    {"read", socket_read},
+    {"write", socket_write},
+    {"setCallback", socket_setCallback},
+    {"setTimeout", socket_setTimeout},
+    {"startTLS", socket_startTLS},
+    {"connected", socket_connected},
+    {"connections", socket_connections},
+    {"info", socket_info},
+
+    {"__tostring", userdata_tostring},
+    {"__gc", userdata_gc},
+
+    {NULL, NULL} // This must end with an empty struct
+};
+
+int luaopen_hs_socket_internal(lua_State *L __unused) {
+    LuaSkin *skin = [LuaSkin shared];
+    refTable = [skin registerLibraryWithObject:USERDATA_TAG
```

Comment:
FWIW, in my testing/refactoring work, I'm moving things to making separate calls to `registerLibrary` and `registerObject`. It's not a big deal at all, but I think it's slightly neater to separate them (there was a time when libraries and their objects shared a metatable, and having a single method for both made more sense). Eventually I'd like to remove `registerLibraryWithObject`.